### PR TITLE
#52 Created CI build workflow to build and push Docker Image

### LINKED ISSUE #52: Create CI build workflow to build and push Docker Image
* Potential Trigger: On pushing a new tag to main branch
* Implementation requires to update file [release_droid_upload_github_release_assets.yml](https://github.com/exasol/data-science-sandbox/blob/main/.github/workflows/release_droid_upload_github_release_assets.yml)
* Note currently the workflow requires AWS codebuild to succeed (see ticket https://github.com/exasol/data-science-sandbox/issues/23) 

### Acceptance criteria 
- [x] Tests have been renamed / moved:
  * `git mv test/aws test/integration/aws`
  * `git mv test/ci test/codebuild` (requires to update script `aws-code-build/ci/buildspec.yaml`, too)
- [x] CI Build has been enhanced to  additionally execute all tests in folder `test/integration`
- [ ] Developer Guide has been enhanced to describe CodeBuild triggering in more detail
- [ ] Generated release letter is enhanced to also name the Docker image

### Details on CodeBuild triggering 

On release 
* user calls release-droid, 
* which in turn runs GitHub action `release_droid_upload_github_release_assets.yml`.
* this calls `exasol.ds.sandbox.main start-release-build`
* `lib/release_build/run_release_build.py`

Currently,  the Webhook is configured in GitHub so that GitHub sends a post request on specific events, e.g. push.
Apparently the event contains the name of the current branch e.g. `refs/heads/main` and the commit message. 
The CodeBuild in the AWS Stack configured to evaluate the event and to start on commit message "[CodeBuild]".

Open questions: 
* The web hooks in the GitHub repo are created by AWS template [templates/ci_code_build.jinja.yaml](https://github.com/exasol/data-science-sandbox/blob/main/exasol/ds/sandbox/templates/ci_code_build.jinja.yaml#L74)
* The ci user used by AWS Codebuild therefore has been granted access to the GitHub repo incl. the permission to create web hooks

### PR DESCRIPTION
**Pull request by ckunki**
closes #52

### CODE REVIEW

File: exasol/ds/sandbox/cli/commands/update_release.py [hunk 1]
```diff
@@ -27,9 +29,15 @@ def update_release(
         asset_id: str,
         log_level: str):
     """
-    This command attaches the links of the release assets (AMI, VM images) to the Github release,
-    indicated by parameter 'release-id'.
+    This command attaches the links of the release assets (AMI, VM images)
+    to the Github release, indicated by parameter 'release-id'.  GitHub token
+    is expected to be found in environment variable GITHUB_TOKEN.
```
Comment (kaklakariada) [inline exasol/ds/sandbox/cli/commands/update_release.py:34]:
```suggestion
    This command attaches the links of the release assets (AMI, VM images)
    to the Github release, indicated by parameter 'release_id'.  GitHub token
    is expected to be found in environment variable 'GITHUB_TOKEN'.
```

File: exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py [hunk 2]
```diff
@@ -28,18 +36,38 @@ def finish(self):
 
 
 class MarkdownTextPrinter(TextPrinter, ABC):
+    def __init__(self, target: io.TextIOBase):
+        self.target = target
+
+    def _format(self, obj: TextObject) -> Optional[str]:
```
Comment (kaklakariada) [inline exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py:42]:
```suggestion
    def _format(self, obj: TextObject) ->str:
```

File: exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py [hunk 3]
```diff
@@ -28,18 +36,38 @@ def finish(self):
...
+        if isinstance(obj, TitleTextObject):
+            return f'#### {obj.text}\n'
+        if isinstance(obj, CodeBlockTextObject):
+            return cleandoc(f"""
+            ```shell
```
Comment (kaklakariada) [inline exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py:47]: If we have different source languages, we might need to add this to `CodeBlockTextObject`

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py [hunk 4]
```diff
@@ -1,33 +1,47 @@
...
-    VM_S3 = "s3-object"
     SNAPSHOT = "snapshot"
     EXPORT_IMAGE_TASK = "export-image-task"
+    VM_S3 = "s3-object"
     CLOUDFORMATION = "cloudformation"
     EC2_KEY_PAIR = "ec2-key-pair"
 
+    @staticmethod
+    def from_name(name: str):
+        return next(a for a in AssetTypes if a.value == name)
```
Comment (kaklakariada) [inline exasol/ds/sandbox/lib/asset_printing/print_assets.py:40]: Check if name is found and throw an exception with a helpful message

File: exasol/ds/sandbox/lib/asset_printing/rich_console_printer.py [hunk 5]
```diff
@@ -26,13 +33,23 @@ def finish(self):
 
 class RichConsoleTextPrinter(TextPrinter, ABC):
 
+    def _style(self, obj: TextObject) -> Optional[str]:
```
Comment (kaklakariada) [inline exasol/ds/sandbox/lib/asset_printing/rich_console_printer.py:36]:
```suggestion
    def _style(self, obj: TextObject) -> Tuple[Optional[str]]:
```

File: exasol/ds/sandbox/lib/aws_access/aws_access.py [hunk 6]
```diff
@@ -100,6 +100,24 @@ def upload_cloudformation_stack(self, yml: str, stack_name: str, tags=tuple()) -
             LOG.error(f"Run 'aws cloudformation describe-stack-events --stack-name {stack_name}' to get details.")
             raise e
 
+    def read_secret_arn(self, physical_resource_id: str):
+        """"
+        Uses Boto3 to retrieve the ARN of a secret.
```
Comment (kaklakariada) [inline exasol/ds/sandbox/lib/aws_access/aws_access.py:105]:
```suggestion
        Uses Boto3 to retrieve the ARN of a secret from secrets manager.
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py (hunk 4 above)
Comment (ckunki) [reply exasol/ds/sandbox/lib/asset_printing/print_assets.py:40]: I will use a dedicated exception and add tests.

File: exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py (hunk 3 above)
Comment (ckunki) [reply exasol/ds/sandbox/lib/asset_printing/mark_down_printer.py:47]: Will do this on demand

File: exasol/ds/sandbox/cli/commands/show_aws_assets.py [hunk 7]
```diff
@@ -35,5 +48,11 @@ def show_aws_assets(
...
     set_log_level(log_level)
     _asset_id = AssetId(asset_id) if asset_id is not None else None
-    print_assets(AwsAccess(aws_profile=aws_profile), asset_id=_asset_id,
-                 outfile=out_file, asset_types=asset_type)
+    asset_types = tuple(AssetTypes.from_name(n) for n in asset_type)
+    with optional_write_to(out_file) as handle:
+        print_assets(
+            AwsAccess(aws_profile=aws_profile),
+            asset_id=_asset_id,
+            outfile=handle,
```
Comment (tkilias) [inline exasol/ds/sandbox/cli/commands/show_aws_assets.py:56]: outfile should be renamed to out_file_obj, but can be done in different PR

File: exasol/ds/sandbox/cli/commands/start_test_release_build.py [hunk 8]
```diff
@@ -9,10 +8,16 @@
...
 from exasol.ds.sandbox.lib.logging import set_log_level
-from exasol.ds.sandbox.lib.github_release_access import GithubReleaseAccess
+from exasol.ds.sandbox.lib.github_release_access import (
+    github_token_or_exit,
+    GithubReleaseAccess,
+)
 from exasol.ds.sandbox.lib.release_build.run_release_build import run_start_test_release_build
 
 
+GITHUB_TOKEN_ENV = "GITHUB_TOKEN"
```
Comment (tkilias) [inline exasol/ds/sandbox/cli/commands/start_test_release_build.py:18]: where is this used?

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py [hunk 9]
```diff
@@ -1,33 +1,47 @@
...
     AMI = "ami"
-    VM_S3 = "s3-object"
     SNAPSHOT = "snapshot"
     EXPORT_IMAGE_TASK = "export-image-task"
+    VM_S3 = "s3-object"
     CLOUDFORMATION = "cloudformation"
     EC2_KEY_PAIR = "ec2-key-pair"
 
+    @staticmethod
+    def from_name(name: str):
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/asset_printing/print_assets.py:39]: Usually, you could use AssetTypes["DOCKER"], but this is only for the real name of the enum and not the assigned string value, but maybe you could overwrite it, but you can also keep it this way

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py [hunk 10]
```diff
@@ -237,41 +251,71 @@ def print_ec2_keys(aws_access: AwsAccess, filter_value: str, printing_factory: P
...
+        AssetTypes.CLOUDFORMATION: print_cloudformation_stacks,
+        AssetTypes.EC2_KEY_PAIR: print_ec2_keys,
+    }
+
+    def second_arg(asset_type: AssetTypes):
+        if asset_type in (AssetTypes.DOCKER, AssetTypes.VM_S3):
+            return asset_id
+        return filter_value
+
+    selected = (a for a in AssetTypes if a in asset_types)
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/asset_printing/print_assets.py:284]:
```suggestion
```

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py [hunk 11]
```diff
@@ -237,41 +251,71 @@ def print_ec2_keys(aws_access: AwsAccess, filter_value: str, printing_factory: P
...
+    for asset_type in selected:
```
Comment (tkilias) [inline exasol/ds/sandbox/lib/asset_printing/print_assets.py:285]:
```suggestion
    for asset_type in asset_types:
```

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol/ds/sandbox/cli/commands/start_test_release_build.py (hunk 8 above)
Comment (ckunki) [reply exasol/ds/sandbox/cli/commands/start_test_release_build.py:18]: Thanks - removed in last push.

File: exasol/ds/sandbox/lib/asset_printing/print_assets.py (hunk 10 above)
Comment (ckunki) [reply exasol/ds/sandbox/lib/asset_printing/print_assets.py:284]: The idea of this line is, to process the values passed in parameter `asset_types` but using the order defined by Enum `AssetTypes`.

File: exasol/ds/sandbox/cli/commands/start_test_release_build.py (hunk 8 above)
Comment (tkilias) [reply exasol/ds/sandbox/cli/commands/start_test_release_build.py:18]: no new commit available, yet

Comment (tkilias) [reply exasol/ds/sandbox/cli/commands/start_test_release_build.py:18]:
```suggestion
```

**Review by tkilias — APPROVED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**